File: 2017/15.cpp
#include <iostream>

constexpr unsigned int modulus = 2147483647;
constexpr unsigned int factorA = 16807;
constexpr unsigned int factorB = 48271;
constexpr unsigned int judge_bits = 0xffff;

/*
unsigned int mulmod_mersenne(
    unsigned long a,
    unsigned int b,
    unsigned int bits,
    unsigned int modulus) {

    const auto v = a * b;
    const auto s = (v & modulus) + (v >> bits);
    return s == modulus ? s - 1 : s;
}
*/

unsigned int mulmod_mersenne(
    unsigned long a,
    unsigned int b,
    unsigned int bits,
    unsigned int modulus) {

    return (a * b) % modulus;
}

unsigned int mulmod_mersenne_even(
    unsigned long a,
    unsigned int b,
    unsigned int bits,
    unsigned int modulus,
    unsigned int low_mask) {

    do {
        a = (a * b) % modulus;
        //std::cout << "v " << v << " lowmask " << low_mask << " and " << (v & low_mask) << std::endl;
        //break;
    } while (a & low_mask);
    return a;
}

int main(int, char**)
{
    unsigned int a = 65;
    unsigned int b = 8921;
    unsigned count = 0;
/*    for (unsigned i = 0; i < 40'000'000; ++i) {
        a = mulmod_mersenne(a, factorA, 31, modulus);
        b = mulmod_mersenne(b, factorB, 31, modulus);
        if ((a & judge_bits) == (b & judge_bits)) {
            ++count;
        }
    } */
    for (unsigned i = 0; i < 5'000'000; ++i) {
        a = mulmod_mersenne_even(a, factorA, 31, modulus, 3);
        //std::cout << "a " << a << std::endl;
        b = mulmod_mersenne_even(b, factorB, 31, modulus, 7);
        //std::cout << "b " << b << std::endl;
        if ((a & judge_bits) == (b & judge_bits)) {
            ++count;
        }
    }
    std::cout << count << std::endl;
}

File: 2017/15.py
#!/usr/bin/python3.12

from pyaoc import Env

e = Env(15)
e.T("""Generator A starts with 65
Generator B starts with 8921""", 588, 309)

factors = [16807, 48271]
modulus = 2147483647
# primitive root 7
judge_bits = 0xffff


def multiply_modulo_mersenne_prime(a, b, n, mod):
    m = a * b
    s = (m >> n) + (m & mod)
    if s == mod:
        s -= 1
    return s

def multiply_modulo(a, b, mod):
    return (a * b) % mod


def part1(input):
    seeds = input.get_ints_tolerant()
    N = 40000000
    #N = 1000
    n = 31
    a = seeds[0]
    b = seeds[1]
    count = 0
    for i in range(N):
        #x = multiply_modulo_mersenne_prime(a, factors[0], n, modulus)
        #y = multiply_modulo_mersenne_prime(b, factors[1], n, modulus)
        a = multiply_modulo(a, factors[0], modulus)
        b = multiply_modulo(b, factors[1], modulus)
        #assert x == a
        #assert y == b
        if (a & judge_bits) == (b & judge_bits):
            count += 1
    return count


e.run_tests(1, part1)
e.run_main(1, part1)


def multiply_modulo_even(a, b, mod, mask):
    while True:
        a = (a * b) % mod
        if (a & mask == 0):
            return a


def part2(input):
    seeds = input.get_ints_tolerant()
    N = 5000000
    # a: v % 4 == 0, b: v % 8 == 0
    n = 31
    a = seeds[0]
    b = seeds[1]
    count = 0
    for i in range(N):
        a = multiply_modulo_even(a, factors[0], modulus, 3)
        b = multiply_modulo_even(b, factors[1], modulus, 7)
        if (a & judge_bits) == (b & judge_bits):
            count += 1
    return count


e.run_tests(2, part2)
e.run_main(2, part2)
